<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>XIBD Introduction • XIBD</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><meta property="og:title" content="XIBD Introduction">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">XIBD</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">XIBD Introduction</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>XIBD Introduction</h1>
                        <h4 class="author">Lyndal Henden</h4>
            
            <h4 class="date">2018-03-19</h4>
          </div>

    
    
<div class="contents">
<p>XIBD performs pairwise relatedness mapping on autosomes and the X chromosome using a first order continuous time hidden Markov model (HMM). First, XIBD estimates the proportion of genome shared identical by descent (IBD) between pairs of samples, which is useful as an initial measure of relatedness. Following this, XIBD can be used to detect genomic regions have been inherited IBD.</p>
<p>The user can select one of two HMMs to implement:</p>
<ol style="list-style-type: decimal">
<li><p><code>model=1</code> is based on the HMM implemented in PLINK (Purcell et al., 2007) which assumes the SNPs are in linkage equilibrium (LE). This often requires thinning of SNPs prior to use.</p></li>
<li><p><code>model=2</code> is based on the HMM implemented in RELATE (Albrechtsen et al., 2009) which allows SNPs to be in LD and implicitly accounts for some (not all) LD through conditional emission probabilities.</p></li>
</ol>
<p>Of these HMMs, <code>model=1</code> has a faster run-time than <code>model=2</code> and is wise to use when then are many SNPs and many samples.</p>
<p>This document outlines the data format and key functions used by XIBD for relatedness mapping. For more details on the algorithm, see:</p>
<p>Henden et al. (2016) XIBD: software for inferring pairwise identity by descent on the X chromosome. <em>Bioinformatics.</em></p>
<p><strong>NOTE: XIBD should not be used with small reference datasets; reference populations that do not match the input population or cohorts of mixed populations.</strong></p>
<div id="data-format" class="section level2">
<h2 class="hasAnchor">
<a href="#data-format" class="anchor"></a>Data Format</h2>
<div id="ped-and-map-files" class="section level3">
<h3 class="hasAnchor">
<a href="#ped-and-map-files" class="anchor"></a>PED and MAP Files</h3>
<p>The input for XIBD is unphased haplotype data for SNPs in PLINK PED and MAP format (<a href="http://www.cog-genomics.org/plink2" class="uri">http://www.cog-genomics.org/plink2</a>).</p>
<p>A PED file is a white-space (space or tab) delimited file with the first six columns</p>
<ol style="list-style-type: decimal">
<li>Family ID</li>
<li>Individual ID</li>
<li>Paternal ID</li>
<li>Maternal ID</li>
<li>Sex (1=male, 2=female)</li>
<li>Phenotype (1=unaffected, 2=affected, 0=unknown)</li>
</ol>
<p>The IDs are alphanumeric: the combination of family and individual ID should uniquely identify a person. Genotypes (column 7 onwards) should also be white-space delimited with the A and B alleles coded as 1 and 2 respectively and missing genotypes coded as 0. All SNPs (whether haploid or not) must have two alleles specified. For haploid chromosomes (X chromosome), genotypes should be specified as homozygous. Either Both alleles should be missing (i.e. 0) or neither. No header row should be given</p>
<p>The MAP file contains exactly 4 columns of information</p>
<ol style="list-style-type: decimal">
<li>Chromosome (1-23, where chromosome 23 is the X chromosome)</li>
<li>rs# or SNP identifier</li>
<li>Genetic map distance (centi-Morgans or Morgans)</li>
<li>Base-pair position</li>
</ol>
<p>where each row describes a single marker. Genetic map distances and base-pair positions are expected to be positive values. The MAP file must be ordered by increasing chromosomes and positions. SNP identifiers can contain any characters except spaces or tabs; also, you should avoid * symbols in names. The MAP file must contain as many markers as are in the PED file. No header row should be given.</p>
</div>
<div id="ld-file" class="section level3">
<h3 class="hasAnchor">
<a href="#ld-file" class="anchor"></a>LD File</h3>
<p>If <code>model=2</code> then a PLINK LD file is also required, see <a href="http://www.cog-genomics.org/plink2" class="uri">http://www.cog-genomics.org/plink2</a> for more details about this file and instructions on how to generate it. The LD file contains linkage disequilibrium information for pairs of SNPs and has 7 columns</p>
<ol style="list-style-type: decimal">
<li>Chromosome of SNP A</li>
<li>Base-pair position of SNP A</li>
<li>rs# or SNP identifier of SNP A</li>
<li>Chromosome of SNP B</li>
<li>Base-pair position of SNP B</li>
<li>rs# or SNP identifier of SNP A</li>
<li>Squared correlation between SNP A and SNP B</li>
</ol>
<p>where each row describes a the correlation between a single pair of SNPs.</p>
</div>
<div id="reference-data" class="section level3">
<h3 class="hasAnchor">
<a href="#reference-data" class="anchor"></a>Reference Data</h3>
<p>XIBD calculates population allele frequencies and haplotype frequencies (<code>model=2</code> only) from either the input dataset or a reference dataset. HapMap phase 2 and 3 PED, MAP and LD reference data (hg19/build 37) for the 11 HapMap populations can be downloaded from <a href="http://bioinf.wehi.edu.au/software/XIBD/" class="uri">http://bioinf.wehi.edu.au/software/XIBD/</a>. These files are in <code>.rds</code> format and can be loaded into R using <code>readRDS</code>.</p>
<p><strong>NOTE: XIBD should not be used with small reference datasets; reference populations that do not match the input population or cohorts of mixed populations.</strong></p>
<table style="width:67%;" class="table">
<colgroup>
<col width="25%">
<col width="41%">
</colgroup>
<thead><tr class="header">
<th>HapMap Annotation</th>
<th>Population Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>ASW</td>
<td>African ancestry on Southwest USA</td>
</tr>
<tr class="even">
<td>CEU</td>
<td>Utah residents with Northern and Western European ancestry from the CEPH collection</td>
</tr>
<tr class="odd">
<td>CHB</td>
<td>Han Chinese in Beijing, China</td>
</tr>
<tr class="even">
<td>CHD</td>
<td>Chinese in Metropolitan Denver, Colorado</td>
</tr>
<tr class="odd">
<td>GIH</td>
<td>Gujarati Indians in Houston, Texas</td>
</tr>
<tr class="even">
<td>JPT</td>
<td>Japanese in Tokyo, Japan</td>
</tr>
<tr class="odd">
<td>LWK</td>
<td>Luhya in Webuye, Kenya</td>
</tr>
<tr class="even">
<td>MEX</td>
<td>Mexican ancestry in Los Angeles, California</td>
</tr>
<tr class="odd">
<td>MKK</td>
<td>Maasai in Kinyawa, Kenya</td>
</tr>
<tr class="even">
<td>TSI</td>
<td>Toscans in Italy</td>
</tr>
<tr class="odd">
<td>YRI</td>
<td>Yoruba in Ibadan, Nigeria (West Africa)</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="key-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#key-functions" class="anchor"></a>Key Functions</h2>
<table style="width:97%;" class="table">
<colgroup>
<col width="26%">
<col width="70%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>calculateAlleleFreq</td>
<td>calculate population allele frequencies</td>
</tr>
<tr class="even">
<td>getGenotypes</td>
<td>get genotypes from haplotype data and perform data filtering</td>
</tr>
<tr class="odd">
<td>getIBDparameters</td>
<td>estimate IBD parameters between pairs</td>
</tr>
<tr class="even">
<td>getIBDsegments</td>
<td>detect IBD segments between pairs</td>
</tr>
<tr class="odd">
<td>getIdentityCoef</td>
<td>calculate identity coefficients from a pedigree</td>
</tr>
<tr class="even">
<td>getLocusMatrix</td>
<td>create a binary matrix of IBD/non-IBD for each SNP and pair</td>
</tr>
<tr class="odd">
<td>getLocusProportions</td>
<td>calculate the proportion of pairs IBD at each SNP</td>
</tr>
<tr class="even">
<td>plotIBDproportions</td>
<td>plot the proportion of pairs IBD across the genome</td>
</tr>
<tr class="odd">
<td>plotIBDsegments</td>
<td>plot the detected IBD segments across the genome</td>
</tr>
<tr class="even">
<td>switchBOTgenotypes</td>
<td>switch A and B alleles for Illumina SNPs named using the TOPBOT method</td>
</tr>
</tbody>
</table>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>Below is an IBD analysis performed on simulated data to demonstrate the steps involved in XIBD. Change the parameters for each function to see how it affects the results.</p>
<p>**NOTE: R has a memory limit and if you try to exceed that limit, the error message begins <code>cannot allocate vector of length</code>. The number of bytes in a character string is limited to 2^31 - 1 ~ 2*10^9, which is also the limit on each dimension of an array.**</p>
<div id="format-and-filter-pedmap" class="section level3">
<h3 class="hasAnchor">
<a href="#format-and-filter-pedmap" class="anchor"></a>Format and Filter PED/MAP</h3>
<p>We begin by formatting and filtering input PED and MAP haplotype.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load XIBD library</span>
<span class="kw">library</span>(XIBD)

<span class="co"># lets look at the data</span>
<span class="kw">str</span>(example_pedmap)</code></pre></div>
<pre><code>## List of 2
##  $ :'data.frame':    10 obs. of  57622 variables:
##   ..$ : num [1:10] 1 1 1 1 1 1 1 1 1 1
##   ..$ : int [1:10] 1 2 4 5 8 9 11 14 15 16
##   ..$ : num [1:10] 0 0 1 1 4 6 7 9 11 13
##   ..$ : num [1:10] 0 0 2 2 3 5 8 10 12 14
##   ..$ : num [1:10] 1 2 1 2 2 1 1 2 2 1
##   ..$ : num [1:10] 1 1 1 1 1 1 1 1 1 1
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 1
##   ..$ : num [1:10] 2 1 2 1 2 1 2 2 2 2
##   ..$ : num [1:10] 1 1 2 2 2 2 1 2 1 1
##   ..$ : num [1:10] 2 1 1 1 1 1 1 2 1 2
##   ..$ : num [1:10] 2 1 1 1 1 2 2 2 2 1
##   ..$ : num [1:10] 1 1 1 1 2 1 2 2 2 2
##   ..$ : num [1:10] 1 1 2 2 2 2 1 2 1 2
##   ..$ : num [1:10] 2 2 1 2 1 2 1 2 1 2
##   ..$ : num [1:10] 2 1 1 1 1 1 1 1 2 2
##   ..$ : num [1:10] 1 1 1 1 2 1 2 1 2 1
##   ..$ : num [1:10] 2 2 2 2 2 1 1 1 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 1 2
##   ..$ : num [1:10] 1 1 1 1 1 1 1 1 1 1
##   ..$ : num [1:10] 1 1 1 1 1 1 1 1 1 1
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 1 1 1 1 2
##   ..$ : num [1:10] 2 1 2 1 1 1 1 1 2 1
##   ..$ : num [1:10] 1 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 1 1 1 1 1 1 1 1 1 1
##   ..$ : num [1:10] 1 1 1 1 1 1 1 1 1 1
##   ..$ : num [1:10] 2 1 1 1 1 2 2 2 2 2
##   ..$ : num [1:10] 1 2 1 2 2 2 2 1 2 1
##   ..$ : num [1:10] 2 1 2 2 2 2 2 2 2 1
##   ..$ : num [1:10] 2 2 1 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 1 2
##   ..$ : num [1:10] 2 1 2 2 2 2 1 2 2 2
##   ..$ : num [1:10] 2 2 1 2 2 2 2 1 2 1
##   ..$ : num [1:10] 1 2 1 1 1 1 1 1 1 1
##   ..$ : num [1:10] 1 1 2 1 1 1 1 1 1 1
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 1 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 1 2
##   ..$ : num [1:10] 2 2 2 2 1 2 1 1 1 1
##   ..$ : num [1:10] 2 1 2 2 2 1 1 1 1 2
##   ..$ : num [1:10] 2 2 1 2 1 2 1 2 1 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 1 1
##   ..$ : num [1:10] 2 2 2 2 1 2 1 2 2 2
##   ..$ : num [1:10] 2 2 1 1 1 2 2 2 2 1
##   ..$ : num [1:10] 1 1 2 1 2 1 2 1 2 1
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 1 1 1 1 1 1 1 1 1 1
##   ..$ : num [1:10] 1 1 1 1 1 1 1 1 2 1
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 1 2
##   ..$ : num [1:10] 2 2 2 2 1 2 1 1 2 1
##   ..$ : num [1:10] 1 1 1 1 1 1 1 1 1 1
##   ..$ : num [1:10] 1 2 1 2 1 2 1 1 1 1
##   ..$ : num [1:10] 1 2 2 2 2 1 1 1 1 2
##   ..$ : num [1:10] 2 1 2 1 1 1 1 1 2 1
##   ..$ : num [1:10] 2 2 1 1 1 2 2 2 2 2
##   ..$ : num [1:10] 1 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 1 1 1 1 1 1 1 1 1 1
##   ..$ : num [1:10] 1 1 1 1 1 1 1 2 1 2
##   ..$ : num [1:10] 2 1 1 1 1 2 2 2 1 2
##   ..$ : num [1:10] 1 2 1 2 1 2 1 1 1 1
##   ..$ : num [1:10] 1 1 1 1 1 1 2 1 2 1
##   ..$ : num [1:10] 1 1 1 1 2 1 2 2 1 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 1 2 2 2 2 1 2 2 1
##   ..$ : num [1:10] 2 2 1 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 2 2 2 1 2 1 2 2
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 1 1 2 2 2 1 2 1 2 2
##   ..$ : num [1:10] 2 1 1 1 2 1 2 2 2 2
##   ..$ : num [1:10] 1 2 1 1 1 1 1 1 1 1
##   ..$ : num [1:10] 1 1 2 1 1 1 1 1 1 1
##   ..$ : num [1:10] 2 2 2 2 2 1 2 1 2 1
##   ..$ : num [1:10] 2 2 2 2 2 2 2 2 1 2
##   ..$ : num [1:10] 1 1 1 1 1 1 1 1 1 2
##   ..$ : num [1:10] 1 2 1 2 1 2 1 1 1 1
##   ..$ : num [1:10] 2 1 2 2 2 2 2 2 1 2
##   ..$ : num [1:10] 2 1 1 1 1 1 1 2 2 2
##   ..$ : num [1:10] 2 1 2 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 2 1 2 2 2 2 2 2 2
##   ..$ : num [1:10] 2 1 1 1 1 1 1 1 1 1
##   ..$ : num [1:10] 1 1 1 1 1 1 1 1 2 1
##   ..$ : num [1:10] 1 1 1 1 1 2 1 2 1 1
##   .. [list output truncated]
##  $ :'data.frame':    28808 obs. of  4 variables:
##   ..$ : int [1:28808] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ : chr [1:28808] "rs7526076" "rs6671356" "rs10907182" "rs4970362" ...
##   ..$ : num [1:28808] 0.0261 0.0267 0.0269 0.0273 0.0281 0.0302 0.0302 0.0303 0.0305 0.0316 ...
##   ..$ : int [1:28808] 998395 1040026 1066403 1094738 1106784 1276077 1287040 1310924 1335218 1404809 ...</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># format and filter the example data</span>
my_genotypes &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getGenotypes.html">getGenotypes</a></span>(<span class="dt">ped.map =</span> example_pedmap,
                             <span class="dt">reference.ped.map =</span> example_reference_pedmap,
                             <span class="dt">snp.ld =</span> example_reference_ld,
                             <span class="dt">model =</span> <span class="dv">2</span>,
                             <span class="dt">maf =</span> <span class="fl">0.01</span>,
                             <span class="dt">sample.max.missing =</span> <span class="fl">0.1</span>,
                             <span class="dt">snp.max.missing =</span> <span class="fl">0.1</span>,
                             <span class="dt">maximum.ld.r2 =</span> <span class="fl">0.99</span>,
                             <span class="dt">chromosomes =</span> <span class="ot">NULL</span>,
                             <span class="dt">input.map.distance =</span> <span class="st">"M"</span>,
                             <span class="dt">reference.map.distance =</span> <span class="st">"M"</span>)</code></pre></div>
<pre><code>## Begin filtering of 10 samples and 28808 SNPs...
## 28808 SNPs remain after merging with reference dataset...
## 28805 SNPs remain after MAF removal...
## 28558 SNPs remain after missingness removal...
## 28297 SNPs remain after LD removal...
## 10 samples remain after missingness removal...</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(my_genotypes)</code></pre></div>
<pre><code>## List of 3
##  $ pedigree :'data.frame':   10 obs. of  6 variables:
##   ..$ fid: num [1:10] 1 1 1 1 1 1 1 1 1 1
##   ..$ iid: int [1:10] 1 2 4 5 8 9 11 14 15 16
##   ..$ pid: num [1:10] 0 0 1 1 4 6 7 9 11 13
##   ..$ mid: num [1:10] 0 0 2 2 3 5 8 10 12 14
##   ..$ sex: num [1:10] 1 2 1 2 2 1 1 2 2 1
##   ..$ aff: num [1:10] 1 1 1 1 1 1 1 1 1 1
##  $ genotypes:'data.frame':   28297 obs. of  21 variables:
##   ..$ chr               : int [1:28297] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ snp_id            : chr [1:28297] "rs7526076" "rs6671356" "rs10907182" "rs4970362" ...
##   ..$ pos_M             : num [1:28297] 0.0261 0.0267 0.0269 0.0273 0.0281 0.0302 0.0302 0.0303 0.0305 0.0316 ...
##   ..$ pos_bp            : int [1:28297] 998395 1040026 1066403 1094738 1106784 1276077 1287040 1310924 1335218 1404809 ...
##   ..$ freq              : num [1:28297] 0.265 0.186 0.58 0.368 0.245 ...
##   ..$ condition_snp     : num [1:28297] -1 1 2 3 4 5 6 7 7 9 ...
##   ..$ pba               : num [1:28297] 0 0.598 0.267 0.136 0.473 ...
##   ..$ pbA               : num [1:28297] 0 0.137 0.543 0.284 0.159 ...
##   ..$ pBa               : num [1:28297] 0 0.216 0.153 0.494 0.281 ...
##   ..$ pBA               : num [1:28297] 0 0.0488 0.0374 0.0857 0.0862 ...
##   ..$ freq_condition_snp: num [1:28297] 0.265 0.265 0.19 0.58 0.368 ...
##   ..$ 1/1               : int [1:28297] 2 1 1 1 1 2 2 2 2 0 ...
##   ..$ 1/2               : int [1:28297] 1 0 0 1 0 2 2 2 2 0 ...
##   ..$ 1/4               : int [1:28297] 2 1 0 1 0 2 2 2 2 0 ...
##   ..$ 1/5               : int [1:28297] 1 1 0 2 0 2 2 2 2 0 ...
##   ..$ 1/8               : int [1:28297] 2 1 1 1 1 2 2 2 2 0 ...
##   ..$ 1/9               : int [1:28297] 1 1 1 2 0 1 2 2 2 0 ...
##   ..$ 1/11              : int [1:28297] 2 0 2 0 1 1 2 2 2 0 ...
##   ..$ 1/14              : int [1:28297] 2 2 2 2 0 1 2 2 2 0 ...
##   ..$ 1/15              : int [1:28297] 2 0 2 0 2 2 2 2 1 0 ...
##   ..$ 1/16              : int [1:28297] 1 1 1 2 1 2 2 2 2 0 ...
##  $ model    : num 2</code></pre>
</div>
<div id="switch-a-and-b-alleles" class="section level3">
<h3 class="hasAnchor">
<a href="#switch-a-and-b-alleles" class="anchor"></a>Switch A and B Alleles</h3>
<p>If the input data has been generated from an Illumina platform <strong>and</strong> HapMap reference data is being used then it is a good idea to check that the A and B alleles as denoted by Illumina match the A and B alleles in the HapMap data.</p>
<p>The HapMap allele frequencies in XIBDs HapMap allele frequency files are calculated for the A allele only, where the A allele is determined by the following rules:</p>
<ol style="list-style-type: decimal">
<li>When one of the possible variations of the SNP is adenine (A), then adenine is labeled the A allele and the remaining variation is labeled the B allele, regardless of what this might be.</li>
<li>If adenine (A) is not a variation of the SNP but cytosine (C) is, then cytosine is labeled the A allele and the remaining variation is labeled the B allele.</li>
<li>If neither adenine (A) or cytosine (C) are variants of the SNP then thymine (T) is labeled the A allele.</li>
</ol>
<p>Illuminas convention for the naming of A and B alleles differs to that of the HapMap data (<a href="http://www.illumina.com/documents/products/technotes/technote_topbot.pdf" class="uri">http://www.illumina.com/documents/products/technotes/technote_topbot.pdf</a>). Rather, the classification of A and B alleles depend on the top (TOP) and bottom (BOT) designations of the SNP. This means that the A allele in the HapMap data is not always the same as the A allele in the Illumina data. In fact, alleles that have been named according to the BOT designation actually correspond to the B allele in the HapMap data. This discrepancy will result in population allele frequencies that differ considerably between the input dataset and the reference dataset, which will produce an unreliable analysis.</p>
<p>To correct for this, <code><a href="../reference/switchBOTgenotypes.html">switchBOTgenotypes()</a></code> switches the A and B alleles in the filtered dataset for all SNPs corresponding to Illumina BOT designations. This mean a homozygous reference genotype, 0, will be changed to a homozygous alternative genotype, 2, and vis versa. Heterozygous genotypes remain unchanged.</p>
<p>We have created an annotation file containing information on the TOP/BOT designations for each SNP in the HapMap data that can be downloaded from <a href="http://bioinf.wehi.edu.au/software/XIBD/" class="uri">http://bioinf.wehi.edu.au/software/XIBD/</a>. This file is required if you need to correct for the TOP/BOT naming convention.</p>
<p>For the purpose of this example, we simulated data using Illumina’s naming convention.</p>
<p><strong>NOTE: this function should only be implemented with Illumina data when HapMap reference data is used and if there is a noticeable discrepancy between population allele frequencies calculated from the HapMap reference data and those calculated from the input dataset.</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># calculate allele frequencies from the input dataset</span>
input_freq &lt;-<span class="st"> </span><span class="kw"><a href="../reference/calculateAlleleFreq.html">calculateAlleleFreq</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes)
<span class="kw">hist</span>(<span class="kw">abs</span>(my_genotypes[[<span class="st">"genotypes"</span>]][,<span class="st">"freq"</span>] <span class="op">-</span><span class="st"> </span>input_freq[,<span class="st">"freq"</span>]),
     <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>), 
     <span class="dt">main =</span> <span class="st">"Before BOT change"</span>,
     <span class="dt">xlab =</span> <span class="st">"abs(pop allele freq diff)"</span>)

<span class="co"># switch alleles</span>
my_genotypes_<span class="dv">2</span> &lt;-<span class="st"> </span><span class="kw"><a href="../reference/switchBOTgenotypes.html">switchBOTgenotypes</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes, 
                                     <span class="dt">hapmap.topbot =</span> example_hapmap_topbot)

<span class="co"># calculate allele frequencies when BOT alleles switched</span>
input_freq &lt;-<span class="st"> </span><span class="kw"><a href="../reference/calculateAlleleFreq.html">calculateAlleleFreq</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes_<span class="dv">2</span>)
<span class="kw">hist</span>(<span class="kw">abs</span>(my_genotypes_<span class="dv">2</span>[[<span class="st">"genotypes"</span>]][,<span class="st">"freq"</span>] <span class="op">-</span><span class="st"> </span>input_freq[,<span class="st">"freq"</span>]),
     <span class="dt">xlim =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>),
     <span class="dt">main =</span> <span class="st">"After BOT change"</span>,
     <span class="dt">xlab =</span> <span class="st">"abs(pop allele freq diff)"</span>)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-2-1.png" width="672"><img src="introduction_files/figure-html/unnamed-chunk-2-2.png" width="672"></p>
</div>
<div id="estimate-parameters" class="section level3">
<h3 class="hasAnchor">
<a href="#estimate-parameters" class="anchor"></a>Estimate Parameters</h3>
<p>Next we estimate the model parameters. These parameters can be useful as an initial measure of relatedness.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># estimate parameters</span>
my_parameters &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getIBDparameters.html">getIBDparameters</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes_<span class="dv">2</span>, 
                                   <span class="dt">number.cores =</span> <span class="dv">1</span>)</code></pre></div>
<pre><code>## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |=                                                                |   2%
  |                                                                       
  |===                                                              |   4%
  |                                                                       
  |====                                                             |   7%
  |                                                                       
  |======                                                           |   9%
  |                                                                       
  |=======                                                          |  11%
  |                                                                       
  |=========                                                        |  13%
  |                                                                       
  |==========                                                       |  16%
  |                                                                       
  |============                                                     |  18%
  |                                                                       
  |=============                                                    |  20%
  |                                                                       
  |==============                                                   |  22%
  |                                                                       
  |================                                                 |  24%
  |                                                                       
  |=================                                                |  27%
  |                                                                       
  |===================                                              |  29%
  |                                                                       
  |====================                                             |  31%
  |                                                                       
  |======================                                           |  33%
  |                                                                       
  |=======================                                          |  36%
  |                                                                       
  |=========================                                        |  38%
  |                                                                       
  |==========================                                       |  40%
  |                                                                       
  |===========================                                      |  42%
  |                                                                       
  |=============================                                    |  44%
  |                                                                       
  |==============================                                   |  47%
  |                                                                       
  |================================                                 |  49%
  |                                                                       
  |=================================                                |  51%
  |                                                                       
  |===================================                              |  53%
  |                                                                       
  |====================================                             |  56%
  |                                                                       
  |======================================                           |  58%
  |                                                                       
  |=======================================                          |  60%
  |                                                                       
  |========================================                         |  62%
  |                                                                       
  |==========================================                       |  64%
  |                                                                       
  |===========================================                      |  67%
  |                                                                       
  |=============================================                    |  69%
  |                                                                       
  |==============================================                   |  71%
  |                                                                       
  |================================================                 |  73%
  |                                                                       
  |=================================================                |  76%
  |                                                                       
  |===================================================              |  78%
  |                                                                       
  |====================================================             |  80%
  |                                                                       
  |=====================================================            |  82%
  |                                                                       
  |=======================================================          |  84%
  |                                                                       
  |========================================================         |  87%
  |                                                                       
  |==========================================================       |  89%
  |                                                                       
  |===========================================================      |  91%
  |                                                                       
  |=============================================================    |  93%
  |                                                                       
  |==============================================================   |  96%
  |                                                                       
  |================================================================ |  98%
  |                                                                       
  |=================================================================| 100%</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(my_parameters)</code></pre></div>
<pre><code>## List of 2
##  $ autosome_parameters    :'data.frame': 45 obs. of  8 variables:
##   ..$ fid1: chr [1:45] "1" "1" "1" "1" ...
##   ..$ iid1: chr [1:45] "1" "1" "1" "1" ...
##   ..$ fid2: chr [1:45] "1" "1" "1" "1" ...
##   ..$ iid2: chr [1:45] "2" "4" "5" "8" ...
##   ..$ m   : num [1:45] 6.06 8.59 1 7.98 10.04 ...
##   ..$ ibd0: num [1:45] 0.97 0 0 0.665 0.65 ...
##   ..$ ibd1: num [1:45] 0.0299 0.9896 1 0.3188 0.3464 ...
##   ..$ ibd2: num [1:45] 0 0.0104 0 0.0158 0.0038 0.0021 0 0.0013 0 0.0079 ...
##  $ X_chromosome_parameters:'data.frame': 45 obs. of  8 variables:
##   ..$ fid1: chr [1:45] "1" "1" "1" "1" ...
##   ..$ iid1: chr [1:45] "1" "1" "1" "1" ...
##   ..$ fid2: chr [1:45] "1" "1" "1" "1" ...
##   ..$ iid2: chr [1:45] "2" "4" "5" "8" ...
##   ..$ m   : num [1:45] 6.41 5.68 1 1 1.9 ...
##   ..$ ibd0: num [1:45] 0.977 0.961 0 0 0.463 ...
##   ..$ ibd1: num [1:45] 0.0235 0.0389 1 1 0.5374 ...
##   ..$ ibd2: num [1:45] 0 0 0 0 0 0 0 0 0 0 ...</code></pre>
</div>
<div id="detect-ibd-segments" class="section level3">
<h3 class="hasAnchor">
<a href="#detect-ibd-segments" class="anchor"></a>Detect IBD Segments</h3>
<p>IBD segments can be detected as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># infer IBD</span>
my_ibd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getIBDsegments.html">getIBDsegments</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes_<span class="dv">2</span>,
                         <span class="dt">parameters =</span> my_parameters, 
                         <span class="dt">model =</span> <span class="ot">NULL</span>, 
                         <span class="dt">chromosomes =</span> <span class="ot">NULL</span>,
                         <span class="dt">number.cores =</span> <span class="dv">1</span>, 
                         <span class="dt">minimum.snps =</span> <span class="dv">20</span>, 
                         <span class="dt">minimum.length.bp =</span> <span class="dv">50000</span>,
                         <span class="dt">error =</span> <span class="fl">0.001</span>, 
                         <span class="dt">posterior =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre><code>## 
  |                                                                       
  |                                                                 |   0%
  |                                                                       
  |=                                                                |   2%
  |                                                                       
  |===                                                              |   4%
  |                                                                       
  |====                                                             |   7%
  |                                                                       
  |======                                                           |   9%
  |                                                                       
  |=======                                                          |  11%
  |                                                                       
  |=========                                                        |  13%
  |                                                                       
  |==========                                                       |  16%
  |                                                                       
  |============                                                     |  18%
  |                                                                       
  |=============                                                    |  20%
  |                                                                       
  |==============                                                   |  22%
  |                                                                       
  |================                                                 |  24%
  |                                                                       
  |=================                                                |  27%
  |                                                                       
  |===================                                              |  29%
  |                                                                       
  |====================                                             |  31%
  |                                                                       
  |======================                                           |  33%
  |                                                                       
  |=======================                                          |  36%
  |                                                                       
  |=========================                                        |  38%
  |                                                                       
  |==========================                                       |  40%
  |                                                                       
  |===========================                                      |  42%
  |                                                                       
  |=============================                                    |  44%
  |                                                                       
  |==============================                                   |  47%
  |                                                                       
  |================================                                 |  49%
  |                                                                       
  |=================================                                |  51%
  |                                                                       
  |===================================                              |  53%
  |                                                                       
  |====================================                             |  56%
  |                                                                       
  |======================================                           |  58%
  |                                                                       
  |=======================================                          |  60%
  |                                                                       
  |========================================                         |  62%
  |                                                                       
  |==========================================                       |  64%
  |                                                                       
  |===========================================                      |  67%
  |                                                                       
  |=============================================                    |  69%
  |                                                                       
  |==============================================                   |  71%
  |                                                                       
  |================================================                 |  73%
  |                                                                       
  |=================================================                |  76%
  |                                                                       
  |===================================================              |  78%
  |                                                                       
  |====================================================             |  80%
  |                                                                       
  |=====================================================            |  82%
  |                                                                       
  |=======================================================          |  84%
  |                                                                       
  |========================================================         |  87%
  |                                                                       
  |==========================================================       |  89%
  |                                                                       
  |===========================================================      |  91%
  |                                                                       
  |=============================================================    |  93%
  |                                                                       
  |==============================================================   |  96%
  |                                                                       
  |================================================================ |  98%
  |                                                                       
  |=================================================================| 100%
## 44 pairs inferred IBD
## 538 IBD segments detected</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">str</span>(my_ibd)</code></pre></div>
<pre><code>## List of 1
##  $ ibd_segments:'data.frame':    538 obs. of  15 variables:
##   ..$ fid1             : chr [1:538] "1" "1" "1" "1" ...
##   ..$ iid1             : chr [1:538] "1" "1" "1" "1" ...
##   ..$ fid2             : chr [1:538] "1" "1" "1" "1" ...
##   ..$ iid2             : chr [1:538] "4" "4" "4" "4" ...
##   ..$ chr              : chr [1:538] "1" "2" "3" "4" ...
##   ..$ start.snp        : chr [1:538] "rs7526076" "rs376385" "rs9682794" "rs4690243" ...
##   ..$ end.snp          : chr [1:538] "rs10788875" "rs4675974" "rs7634881" "rs2739584" ...
##   ..$ start.position.bp: num [1:538] 998395 190245 70895 215351 442790 ...
##   ..$ end.position.bp  : num [1:538] 2.49e+08 2.42e+08 1.98e+08 1.91e+08 1.80e+08 ...
##   ..$ start.position.M : num [1:538] 0.0261 0.0007 0.0001 0.0037 0.0032 0.0032 0.0033 0.0002 0.0019 0.0004 ...
##   ..$ end.position.M   : num [1:538] 2.86 2.68 2.23 2.15 2.03 ...
##   ..$ number.snps      : num [1:538] 2310 2375 1774 1569 1859 ...
##   ..$ length.bp        : num [1:538] 2.48e+08 2.42e+08 1.98e+08 1.91e+08 1.80e+08 ...
##   ..$ length.M         : num [1:538] 2.84 2.68 2.23 2.14 2.03 ...
##   ..$ ibd.status       : num [1:538] 1 1 1 1 1 1 1 1 1 1 ...</code></pre>
</div>
<div id="plotting-results" class="section level3">
<h3 class="hasAnchor">
<a href="#plotting-results" class="anchor"></a>Plotting Results</h3>
<p>XIBD comes with several plotting functions to make interpretation of IBD results easier. Below we plot detected IBD segments across the genome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># plot IBD segments</span>
<span class="kw"><a href="../reference/plotIBDsegments.html">plotIBDsegments</a></span>(<span class="dt">ibd.segments =</span> my_ibd, 
                <span class="dt">ped.genotypes =</span> my_genotypes_<span class="dv">2</span>, 
                <span class="dt">interval =</span> <span class="ot">NULL</span>,
                <span class="dt">annotation.genes =</span> <span class="ot">NULL</span>, 
                <span class="dt">highlight.genes =</span> <span class="ot">NULL</span>, 
                <span class="dt">segment.height =</span> <span class="fl">0.5</span>,
                <span class="dt">number.per.page =</span> <span class="ot">NULL</span>, 
                <span class="dt">add.fid.name =</span> <span class="ot">FALSE</span>, 
                <span class="dt">add.iid.name =</span> <span class="ot">TRUE</span>,
                <span class="dt">add.rug =</span> <span class="ot">TRUE</span>, 
                <span class="dt">plot.title =</span> <span class="st">"Inferred IBD Segments"</span>, 
                <span class="dt">add.legend =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
<p>We can also look to investigate if some regions of the genome have more IBD than others by plotting the proportion of pairs IBD at each SNP.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get IBD proportions</span>
my_locus_matrix &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getLocusMatrix.html">getLocusMatrix</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes_<span class="dv">2</span>, 
                               <span class="dt">ibd.segments =</span> my_ibd)
my_locus_prop &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getLocusProportion.html">getLocusProportion</a></span>(<span class="dt">ped.genotypes =</span> my_genotypes_<span class="dv">2</span>, 
                                 <span class="dt">locus.matrix =</span> my_locus_matrix, 
                                 <span class="dt">groups =</span> <span class="ot">NULL</span>)

<span class="co"># plot IBD proportions</span>
<span class="kw"><a href="../reference/plotIBDproportions.html">plotIBDproportions</a></span>(<span class="dt">locus.proportions =</span> my_locus_prop, 
                   <span class="dt">interval =</span> <span class="ot">NULL</span>,
                   <span class="dt">annotation.genes =</span> <span class="ot">NULL</span>, 
                   <span class="dt">highlight.genes =</span> <span class="ot">NULL</span>, 
                   <span class="dt">add.rug =</span> <span class="ot">TRUE</span>,
                   <span class="dt">plot.title =</span> <span class="ot">NULL</span>)</code></pre></div>
<p><img src="introduction_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
</div>
<div id="identity-coefficients" class="section level3">
<h3 class="hasAnchor">
<a href="#identity-coefficients" class="anchor"></a>Identity Coefficients</h3>
<p>Identity coefficients can be computed given a pedigree as follows.</p>
<p><strong>NOTE: this function requires all individual IDs to be numeric and hence some manual formatting of the pedigree may be required. Additionally, individuals in the pedigree should be numbered in a way such that every parent precedes his or her children.</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get identity coefficients</span>
my_pedigree &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">fid =</span> <span class="kw">rep</span>(<span class="dv">1</span>,<span class="dv">16</span>), 
                          <span class="dt">iid =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">16</span>, 
                          <span class="dt">pid =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">0</span>,<span class="dv">7</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">9</span>,<span class="dv">11</span>,<span class="dv">13</span>),
                          <span class="dt">mid =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">3</span>,<span class="dv">5</span>,<span class="dv">0</span>,<span class="dv">8</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">10</span>,<span class="dv">12</span>,<span class="dv">14</span>), 
                          <span class="dt">sex =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">2</span>,<span class="dv">1</span>), 
                          <span class="dt">aff =</span> <span class="kw">rep</span>(<span class="dv">1</span>,<span class="dv">16</span>))

identity_coef &lt;-<span class="st"> </span><span class="kw"><a href="../reference/getIdentityCoef.html">getIdentityCoef</a></span>(<span class="dt">pedigree =</span> my_pedigree, 
                                 <span class="dt">number.cores =</span> <span class="dv">1</span>)

<span class="kw">head</span>(identity_coef)</code></pre></div>
<pre><code>##   fid1 iid1 fid2 iid2 a_i0 a_i1 a_i2 x_i0 x_i1 x_i2
## 1    1    1    1    2    1    0    0    1    0    0
## 2    1    1    1    3    1    0    0    1    0    0
## 3    1    1    1    4    0    1    0    1    0    0
## 4    1    1    1    5    0    1    0    0    1    0
## 5    1    1    1    6    1    0    0    1    0    0
## 6    1    1    1    7    1    0    0    1    0    0</code></pre>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#data-format">Data Format</a></li>
      <li><a href="#key-functions">Key Functions</a></li>
      <li><a href="#example">Example</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Lyndal Henden.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
