# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

IBDMatrix <- function(chromosomes, positions_bp, number_pairs, ibd_pairs_colnumbers, ibd_chromosomes, ibd_start_bp, ibd_stop_bp) {
    .Call('_XIBD_IBDMatrix', PACKAGE = 'XIBD', chromosomes, positions_bp, number_pairs, ibd_pairs_colnumbers, ibd_chromosomes, ibd_start_bp, ibd_stop_bp)
}

calculateMissingness <- function(genotypes) {
    .Call('_XIBD_calculateMissingness', PACKAGE = 'XIBD', genotypes)
}

calculatePopAlleleFreq <- function(genotypes, chromosomes, genders) {
    .Call('_XIBD_calculatePopAlleleFreq', PACKAGE = 'XIBD', genotypes, chromosomes, genders)
}

groupPairs <- function(group) {
    .Call('_XIBD_groupPairs', PACKAGE = 'XIBD', group)
}

calculateHaplotypeFreq <- function(genotypes, condition_snps_0, genders, chromosome, polyrootR) {
    .Call('_XIBD_calculateHaplotypeFreq', PACKAGE = 'XIBD', genotypes, condition_snps_0, genders, chromosome, polyrootR)
}

haplotypeToGenotype <- function(haplotypes, chromosomes, genders) {
    .Call('_XIBD_haplotypeToGenotype', PACKAGE = 'XIBD', haplotypes, chromosomes, genders)
}

IBDLabel <- function(snp_id, number_snps) {
    .Call('_XIBD_IBDLabel', PACKAGE = 'XIBD', snp_id, number_snps)
}

isolatePairs <- function(fid, iid) {
    .Call('_XIBD_isolatePairs', PACKAGE = 'XIBD', fid, iid)
}

calculate_viterbi_m1 <- function(noStates, piProb, meiosis, NoSNPs, genotypes, alleleFreq, positionM, error, gender1, gender2, chrom) {
    .Call('_XIBD_calculate_viterbi_m1', PACKAGE = 'XIBD', noStates, piProb, meiosis, NoSNPs, genotypes, alleleFreq, positionM, error, gender1, gender2, chrom)
}

calculate_gamma_m1 <- function(noStates, piProb, meiosis, NoSNPs, genotypes, alleleFreq, positionM, error, gender1, gender2, chrom) {
    .Call('_XIBD_calculate_gamma_m1', PACKAGE = 'XIBD', noStates, piProb, meiosis, NoSNPs, genotypes, alleleFreq, positionM, error, gender1, gender2, chrom)
}

calculate_viterbi_m2 <- function(noStates, piProb, meiosis, NoSNPs, genotypes, alleleFreq, positionM, error, gender1, gender2, chrom, conditionSNPs, haplotypeFreq) {
    .Call('_XIBD_calculate_viterbi_m2', PACKAGE = 'XIBD', noStates, piProb, meiosis, NoSNPs, genotypes, alleleFreq, positionM, error, gender1, gender2, chrom, conditionSNPs, haplotypeFreq)
}

calculate_gamma_m2 <- function(noStates, piProb, meiosis, NoSNPs, genotypes, alleleFreq, positionM, error, gender1, gender2, chrom, conditionSNPs, haplotypeFreq) {
    .Call('_XIBD_calculate_gamma_m2', PACKAGE = 'XIBD', noStates, piProb, meiosis, NoSNPs, genotypes, alleleFreq, positionM, error, gender1, gender2, chrom, conditionSNPs, haplotypeFreq)
}

bVectorHH <- function(genotypes) {
    .Call('_XIBD_bVectorHH', PACKAGE = 'XIBD', genotypes)
}

AmatrixHH <- function(pop_allele_freqs, genotypes) {
    .Call('_XIBD_AmatrixHH', PACKAGE = 'XIBD', pop_allele_freqs, genotypes)
}

bVectorHD <- function(genotypes) {
    .Call('_XIBD_bVectorHD', PACKAGE = 'XIBD', genotypes)
}

AmatrixHD <- function(pop_allele_freqs, genotypes) {
    .Call('_XIBD_AmatrixHD', PACKAGE = 'XIBD', pop_allele_freqs, genotypes)
}

bVectorDD <- function(genotypes) {
    .Call('_XIBD_bVectorDD', PACKAGE = 'XIBD', genotypes)
}

AmatrixDD <- function(pop_allele_freqs, genotypes) {
    .Call('_XIBD_AmatrixDD', PACKAGE = 'XIBD', pop_allele_freqs, genotypes)
}

topbotChange <- function(Genotypes, topbot) {
    .Call('_XIBD_topbotChange', PACKAGE = 'XIBD', Genotypes, topbot)
}

