<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>XIBDs IBD Segment Detection — getIBDsegments • XIBD</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<meta property="og:title" content="XIBDs IBD Segment Detection — getIBDsegments" />

<meta property="og:description" content="Detects genomic regions shared IBD between pairs." />
<meta name="twitter:card" content="summary" />
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">XIBD</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction.html">XIBD Introduction</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>XIBDs IBD Segment Detection</h1>
    </div>

    
    <p>Detects genomic regions shared IBD between pairs.</p>
    

    <pre class="usage"><span class='fu'>getIBDsegments</span>(<span class='no'>ped.genotypes</span>, <span class='no'>parameters</span>, <span class='kw'>model</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>chromosomes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>number.cores</span> <span class='kw'>=</span> <span class='fl'>1</span>, <span class='kw'>minimum.snps</span> <span class='kw'>=</span> <span class='fl'>20</span>, <span class='kw'>minimum.length.bp</span> <span class='kw'>=</span> <span class='fl'>50000</span>,
  <span class='kw'>error</span> <span class='kw'>=</span> <span class='fl'>0.001</span>, <span class='kw'>posterior</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ped.genotypes</th>
      <td><p>a named list containing <code>pedigree</code>, <code>genotypes</code> and <code>model</code>.
See <code>Value</code> description in <code><a href='getGenotypes.html'>getGenotypes</a></code> for more details.
Note the family IDs and individual IDs in <code>pedigree</code> must match the family IDs and individual IDs in the header of <code>genotypes</code>.</p></td>
    </tr>
    <tr>
      <th>parameters</th>
      <td><p>a data frame containing meioses and IBD probability estimates for all pairwise combinations of samples.
See <code>Value</code> description in <code><a href='getIBDparameters.html'>getIBDparameters</a></code> for more details.</p></td>
    </tr>
    <tr>
      <th>model</th>
      <td><p>an integer of either 1 or 2 denoting which of the two models should be run.</p><ol>
<li><p><code>model=1</code> is based on the HMM implemented in PLINK (Purcell et al., 2007) which assumes the SNPs
are in linkage equilibrium (LE). This often requires thinning of markers prior to use.</p></li>
<li><p><code>model=2</code> is based on the HMM implemented in RELATE (Albrechtsen et al., 2009) which allows SNPs
to be in LD and implicitly accounts for the LD through conditional emission probabilities where
the current genotype probability is conditioned on the genotype of a single previous SNP.</p></li>
</ol><p>The default is <code>model=NULL</code> and the model listed in <code>ped.genotypes</code> is used. NOTE: <code>model=2</code> can
only be used if <code>model=2</code> is in <code>ped.genotypes</code>.</p></td>
    </tr>
    <tr>
      <th>chromosomes</th>
      <td><p>a numeric vector containing a subset of chromosomes to perform IBD analysis on. The
default is <code>chromosomes=NULL</code> and IBD analysis will be performed on all chromosomes in <code>ped.genotypes</code>.</p></td>
    </tr>
    <tr>
      <th>number.cores</th>
      <td><p>the number of cores used for parallel execution.</p></td>
    </tr>
    <tr>
      <th>minimum.snps</th>
      <td><p>the minimum number of SNPs in an IBD segment for it to be reported. The default value is 20 SNPs.</p></td>
    </tr>
    <tr>
      <th>minimum.length.bp</th>
      <td><p>the minimum length of a reported IBD segment. The default value is 50,000 bp.</p></td>
    </tr>
    <tr>
      <th>error</th>
      <td><p>the genotyping error rate. The default value is 0.001.</p></td>
    </tr>
    <tr>
      <th>posterior</th>
      <td><p>a logical value indicating whether posterior probabilities for each pairwise analysis should be returned.
The posterior probability is calculated for each SNP as posteriorPr(IBD=1)/2 + posteriorPr(IBD=2) using the
forward and backward variables (Rabiner, 1989).
A data frame containing probabilities for each SNP and each pairwise analysis is returned.
This data frame can be very large when there are many SNPs and many pairwise analyses, and the run-time of
<code>getIBDsegments()</code> will increase. The default is <code>posterior=FALSE</code>; <code>posterior=TRUE</code> is not recommended for large datasets.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A named list of 1 object when <code>posterior=FALSE</code> and 2 objects when <code>posterior=TRUE</code>.
The first object in the list, <code>ibd_segments</code>, is a data frame with information:</p><ol>
<li><p>Family 1 ID (type <code>"character"</code>)</p></li>
<li><p>Individual 1 ID (type <code>"character"</code>)</p></li>
<li><p>Family 2 ID (type <code>"character"</code>)</p></li>
<li><p>Individual 2 ID (type <code>"character"</code>)</p></li>
<li><p>Chromosome (type <code>"numeric"</code> or <code>integer</code>)</p></li>
<li><p>SNP identifier (type <code>"character"</code>)</p></li>
<li><p>Start SNP (type <code>"character"</code>)</p></li>
<li><p>End SNP (type <code>"character"</code>)</p></li>
<li><p>Start position bp (type <code>"numeric"</code> or <code>integer</code>)</p></li>
<li><p>End position bp (type <code>"numeric"</code> or <code>integer</code>)</p></li>
<li><p>Start position M (type <code>"numeric"</code>)</p></li>
<li><p>End position M (type <code>"numeric"</code>)</p></li>
<li><p>Number of SNPs (type <code>"numeric"</code> or <code>integer</code>)</p></li>
<li><p>Length bp (type <code>"numeric"</code> or <code>integer</code>)</p></li>
<li><p>Length M (type <code>"numeric"</code>)</p></li>
<li><p>IBD status (1 = one allele shared IBD, 2 = two alleles shared IBD) (type <code>"numeric"</code> or <code>integer</code>)</p></li>
</ol><p>where each row is a unique IBD segment for a pair of individuals. The data frame is headed
<code>fid1, iid1, fid2, iid2, chr, start.snp, end.snp, start.position.bp, end.position.bp, start.position.M, end.position.M,
number.snps, length.bp, length.M, ibd.status</code>. The second object (returned when <code>posterior=TRUE</code>), <code>posterior_probabilities</code>, is a
data frame with the first four columns</p><ol>
<li><p>Chromosome (type <code>"numeric"</code> or <code>integer</code>)</p></li>
<li><p>SNP identifier (type <code>"character"</code>)</p></li>
<li><p>Genetic map distance (type <code>"numeric"</code>)</p></li>
<li><p>Base-pair positions (type <code>"numeric"</code> or <code>integer</code>)</p></li>
</ol><p>and columns 5 onwards are the posterior probabilities for each pair with pair identifier headers.
Rows correspond to SNPs.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='co'># infer IBD</span>
<span class='no'>my_ibd</span> <span class='kw'>&lt;-</span> <span class='fu'>getIBDsegments</span>(<span class='kw'>ped.genotypes</span> <span class='kw'>=</span> <span class='no'>example_genotypes</span>,
                         <span class='kw'>parameters</span> <span class='kw'>=</span> <span class='no'>example_parameters</span>,
                         <span class='kw'>model</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
                         <span class='kw'>chromosomes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
                         <span class='kw'>number.cores</span> <span class='kw'>=</span> <span class='fl'>1</span>,
                         <span class='kw'>minimum.snps</span> <span class='kw'>=</span> <span class='fl'>20</span>,
                         <span class='kw'>minimum.length.bp</span> <span class='kw'>=</span> <span class='fl'>50000</span>,
                         <span class='kw'>error</span> <span class='kw'>=</span> <span class='fl'>0.001</span>,
                         <span class='kw'>posterior</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>)

<span class='fu'>str</span>(<span class='no'>my_ibd</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Lyndal Henden.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>

