<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>XIBD Locus Proportion Plot — plotIBDproportions • XIBD</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<meta property="og:title" content="XIBD Locus Proportion Plot — plotIBDproportions" />

<meta property="og:description" content="Plot the proportion of pairs IBD for each SNP across the genome.
Annotation genes can be added to the plot and specific genes on interest can be highlighted." />
<meta name="twitter:card" content="summary" />
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">XIBD</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction.html">XIBD Introduction</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>XIBD Locus Proportion Plot</h1>
    </div>

    
    <p>Plot the proportion of pairs IBD for each SNP across the genome.
Annotation genes can be added to the plot and specific genes on interest can be highlighted.</p>
    

    <pre class="usage"><span class='fu'>plotIBDproportions</span>(<span class='no'>locus.proportions</span>, <span class='kw'>interval</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>annotation.genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>highlight.genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>add.rug</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
  <span class='kw'>plot.title</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>locus.proportions</th>
      <td><p>a data frame containing the proportion of pairs IBD at each SNP.
See <code>value</code> description in <code><a href='getLocusProportion.html'>getLocusProportion</a></code> for more details.</p></td>
    </tr>
    <tr>
      <th>interval</th>
      <td><p>a vector of length 3 containing a chromosome, a start position (bp) and an end position (bp)
of an interval to plot. The default is <code>interval=NULL</code> which will plot the proportions over all chromosomes
in <code>locus.proportions</code>.</p></td>
    </tr>
    <tr>
      <th>annotation.genes</th>
      <td><p>a data frame with at least 5 columns of information:</p><ol>
<li><p>Chromosome (type <code>"numeric"</code> or <code>"integer"</code>)</p></li>
<li><p>Gene name (type <code>"character"</code>)</p></li>
<li><p>Start location of the gene in base-pairs (type <code>"numeric"</code> or <code>"integer"</code>)</p></li>
<li><p>End location of the gene in base-pairs (type <code>"numeric"</code> or <code>"integer"</code>)</p></li>
<li><p>Gene strand (+ or -)</p></li>
</ol><p><code>annotation.genes</code> should contain the following headers <code>chr, name, start, end</code> and <code>strand</code>.
This data frame does not have to be in a specific order, however it must contain all of the above information
with respective labels.</p></td>
    </tr>
    <tr>
      <th>highlight.genes</th>
      <td><p>a data frame with at least 4 columns of information:</p><ol>
<li><p>Chromosome (type <code>"numeric"</code> or <code>"integer"</code>)</p></li>
<li><p>Gene name (type <code>"character"</code>)</p></li>
<li><p>Start location of the gene in base-pairs (type <code>"numeric"</code> or <code>"integer"</code>)</p></li>
<li><p>End location of the gene in base-pairs (type <code>"numeric"</code> or <code>"integer"</code>)</p></li>
</ol><p><code>highlight.genes</code> should contain the following headers <code>chr, name, start</code> and <code>end</code>.
This data frame does not have to be in a specific order, however it must contain all of the above information
with respective labels.</p></td>
    </tr>
    <tr>
      <th>add.rug</th>
      <td><p>Logical. Whether to include SNP positions as a rug in the figure. The default is <code>add.rug=FALSE</code></p></td>
    </tr>
    <tr>
      <th>plot.title</th>
      <td><p>a character string of a title to be added to the plot. The default is <code>plot.title=NULL</code> which does not add a title to the plot.</p></td>
    </tr>
    </table>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># generate a binary IBD matrix</span>
<span class='no'>my_locus_matrix</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='getLocusMatrix.html'>getLocusMatrix</a></span>(<span class='kw'>ped.genotypes</span> <span class='kw'>=</span> <span class='no'>example_genotypes</span>,
                                  <span class='kw'>ibd.segments</span> <span class='kw'>=</span> <span class='no'>example_ibd</span>)

<span class='co'># calculate the proportion of pairs IBD at each SNP</span>
<span class='no'>my_locus_prop</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='getLocusProportion.html'>getLocusProportion</a></span>(<span class='kw'>ped.genotypes</span> <span class='kw'>=</span> <span class='no'>example_genotypes</span>,
                                    <span class='kw'>locus.matrix</span> <span class='kw'>=</span> <span class='no'>my_locus_matrix</span>,
                                    <span class='kw'>groups</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)

<span class='co'># plot the proportion of pairs IBD</span>
<span class='fu'>plotIBDproportions</span>(<span class='kw'>locus.proportions</span> <span class='kw'>=</span> <span class='no'>my_locus_prop</span>,
                   <span class='kw'>interval</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
                   <span class='kw'>annotation.genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
                   <span class='kw'>highlight.genes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
                   <span class='kw'>add.rug</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
                   <span class='kw'>plot.title</span> <span class='kw'>=</span> <span class='kw'>NULL</span>)</div><div class='img'><img src='plotIBDproportions-1.png' alt='' width='700' height='432.632880098887' /></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
            
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Lyndal Henden.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>

