<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>XIBD BOT Genotype Switching — switchBOTgenotypes • XIBD</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<meta property="og:title" content="XIBD BOT Genotype Switching — switchBOTgenotypes" />

<meta property="og:description" content="The HapMap allele frequencies in XIBDs HapMap allele frequency files are calculated for the A allele only,
where the A allele is determined by the following rules:
When one of the possible variations of the SNP is adenine (A), then adenine is labeled the A allele
and the remaining variation is labeled the B allele, regardless of what this might be.
If adenine (A) is not a variation of the SNP but cytosine (C) is, then cytosine is labeled the A allele
and the remaining variation is labeled the B allele.
If neither adenine (A) or cytosine (C) are variants of the SNP then thymine (T) is labeled the A allele.
Illuminas convention for the naming of A and B alleles differs to that of the HapMap data
(http://www.illumina.com/documents/products/technotes/technote_topbot.pdf). Rather, the classification
of A and B alleles depend on the top (TOP) and bottom (BOT) designations of the SNP. This
means that the A allele in the HapMap data is not always the same as the A allele in the Illumina data. In
fact, alleles that have been named according to the BOT designation actually correspond the the B allele
in the HapMap data. To correct for this, switchBOTgenotypes() switchs the A and B alleles in
the input genotypes for all SNPs corresponding to BOT designations. This mean a homozygous genotype, 0, will be
changed to a homozygous alternative genotype, 2, and vis versa. Heterozygous genotypes will be unchanged.
NOTE: this function should only be implemented with Illumina SNPchip data when XIBD's HapMap reference data is used
and if there is a noticeable discrepancy between population allele frequencies calculated from the HapMap reference data
and those calculated from the input dataset." />
<meta name="twitter:card" content="summary" />
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">XIBD</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/introduction.html">XIBD Introduction</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>XIBD BOT Genotype Switching</h1>
    </div>

    
    <p>The HapMap allele frequencies in XIBDs HapMap allele frequency files are calculated for the A allele only,
where the A allele is determined by the following rules:</p><ol>
<li><p>When one of the possible variations of the SNP is adenine (A), then adenine is labeled the A allele
and the remaining variation is labeled the B allele, regardless of what this might be.</p></li>
<li><p>If adenine (A) is not a variation of the SNP but cytosine (C) is, then cytosine is labeled the A allele
and the remaining variation is labeled the B allele.</p></li>
<li><p>If neither adenine (A) or cytosine (C) are variants of the SNP then thymine (T) is labeled the A allele.</p></li>
</ol><p>Illuminas convention for the naming of A and B alleles differs to that of the HapMap data
(<a href='http://www.illumina.com/documents/products/technotes/technote_topbot.pdf'>http://www.illumina.com/documents/products/technotes/technote_topbot.pdf</a>). Rather, the classification
of A and B alleles depend on the top (TOP) and bottom (BOT) designations of the SNP. This
means that the A allele in the HapMap data is not always the same as the A allele in the Illumina data. In
fact, alleles that have been named according to the BOT designation actually correspond the the B allele
in the HapMap data. To correct for this, <code>switchBOTgenotypes()</code> switchs the A and B alleles in
the input genotypes for all SNPs corresponding to BOT designations. This mean a homozygous genotype, 0, will be
changed to a homozygous alternative genotype, 2, and vis versa. Heterozygous genotypes will be unchanged.
NOTE: this function should only be implemented with Illumina SNPchip data when XIBD's HapMap reference data is used
and if there is a noticeable discrepancy between population allele frequencies calculated from the HapMap reference data
and those calculated from the input dataset.</p>
    

    <pre class="usage"><span class='fu'>switchBOTgenotypes</span>(<span class='no'>ped.genotypes</span>, <span class='no'>hapmap.topbot</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>ped.genotypes</th>
      <td><p>a named list containing <code>pedigree</code>, <code>genotypes</code> and <code>model</code>.
See <code>Value</code> description in <code><a href='getGenotypes.html'>getGenotypes</a></code> for more details.
The family IDs and individual IDs in <code>pedigree</code> must match the family IDs and individual IDs in the header of <code>genotypes</code>.</p></td>
    </tr>
    <tr>
      <th>hapmap.topbot</th>
      <td><p>a data frame containing the Illumina TOP/BOT designation for the HapMap SNPs.
This file can be downloaded from <a href='http://bioinf.wehi.edu.au/software/XIBD/index.html'>http://bioinf.wehi.edu.au/software/XIBD/index.html</a>.
This file contains the following 7 columns of information:</p><ol>
<li><p>Chromosome (<code>"numeric"</code> or <code>"integer"</code>)</p></li>
<li><p>SNP identifier (type <code>"character"</code>)</p></li>
<li><p>Genetic map distance (centi morgans cM, or morgans M - default) (type <code>"numeric"</code>)</p></li>
<li><p>Base-pair position (type <code>"numeric"</code> or <code>"integer"</code>)</p></li>
<li><p>Illuminas TOP or BOT designation of the SNP (type <code>"character"</code>)</p></li>
</ol><p>where each row describes a single marker. The data frame should contain the header
<code>chr, snp_id, pos_bp, pos_M</code> and <code>TOPBOT</code>.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A named list of the same format as the input <code>ped.genotypes</code> with A and B alleles switched for BOT SNPs.</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># The following should only be run if you have Illumina data and</span>
<span class='co'># are using the HapMap reference data provided by XIBD.</span>

<span class='co'># format and filter the data</span>
<span class='no'>my_genotypes</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='getGenotypes.html'>getGenotypes</a></span>(<span class='kw'>ped.map</span> <span class='kw'>=</span> <span class='no'>example_pedmap</span>,
                             <span class='kw'>reference.ped.map</span> <span class='kw'>=</span> <span class='no'>example_reference_pedmap</span>,
                             <span class='kw'>snp.ld</span> <span class='kw'>=</span> <span class='no'>example_reference_ld</span>,
                             <span class='kw'>model</span> <span class='kw'>=</span> <span class='fl'>2</span>,
                             <span class='kw'>maf</span> <span class='kw'>=</span> <span class='fl'>0.01</span>,
                             <span class='kw'>sample.max.missing</span> <span class='kw'>=</span> <span class='fl'>0.1</span>,
                             <span class='kw'>snp.max.missing</span> <span class='kw'>=</span> <span class='fl'>0.1</span>,
                             <span class='kw'>maximum.ld.r2</span> <span class='kw'>=</span> <span class='fl'>0.99</span>,
                             <span class='kw'>chromosomes</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
                             <span class='kw'>input.map.distance</span> <span class='kw'>=</span> <span class='st'>"M"</span>,
                             <span class='kw'>reference.map.distance</span> <span class='kw'>=</span> <span class='st'>"M"</span>)</div><div class='output co'>#&gt; Begin filtering of 10 samples and 28808 SNPs...
#&gt; 28808 SNPs remain after merging with reference dataset...
#&gt; 28805 SNPs remain after MAF removal...
#&gt; 28558 SNPs remain after missingness removal...
#&gt; 28297 SNPs remain after LD removal...
#&gt; 10 samples remain after missingness removal...</div><div class='input'>
<span class='co'># calculate allele frequencies from the input dataset</span>
<span class='no'>input_freq</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='calculateAlleleFreq.html'>calculateAlleleFreq</a></span>(<span class='kw'>ped.genotypes</span> <span class='kw'>=</span> <span class='no'>my_genotypes</span>)
<span class='fu'>hist</span>(<span class='fu'>abs</span>(<span class='no'>my_genotypes</span><span class='kw'>[[</span><span class='st'>"genotypes"</span>]][,<span class='st'>"freq"</span>] - <span class='no'>input_freq</span>[,<span class='st'>"freq"</span>]),
     <span class='kw'>xlim</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>0</span>,<span class='fl'>1</span>),
     <span class='kw'>main</span> <span class='kw'>=</span> <span class='st'>"Before BOT change"</span>,
     <span class='kw'>xlab</span> <span class='kw'>=</span> <span class='st'>"abs(pop allele freq diff)"</span>)</div><div class='img'><img src='switchBOTgenotypes-1.png' alt='' width='700' height='432.632880098887' /></div><div class='input'>
<span class='co'># switch alleles</span>
<span class='no'>my_genotypes_2</span> <span class='kw'>&lt;-</span> <span class='fu'>switchBOTgenotypes</span>(<span class='kw'>ped.genotypes</span> <span class='kw'>=</span> <span class='no'>my_genotypes</span>,
                                     <span class='kw'>hapmap.topbot</span> <span class='kw'>=</span> <span class='no'>example_hapmap_topbot</span>)

<span class='co'># calculate allele frequencies when BOT alleles switched</span>
<span class='no'>input_freq</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='calculateAlleleFreq.html'>calculateAlleleFreq</a></span>(<span class='kw'>ped.genotypes</span> <span class='kw'>=</span> <span class='no'>my_genotypes_2</span>)
<span class='fu'>hist</span>(<span class='fu'>abs</span>(<span class='no'>my_genotypes_2</span><span class='kw'>[[</span><span class='st'>"genotypes"</span>]][,<span class='st'>"freq"</span>] - <span class='no'>input_freq</span>[,<span class='st'>"freq"</span>]),
     <span class='kw'>xlim</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='fl'>0</span>,<span class='fl'>1</span>),
     <span class='kw'>main</span> <span class='kw'>=</span> <span class='st'>"After BOT change"</span>,
     <span class='kw'>xlab</span> <span class='kw'>=</span> <span class='st'>"abs(pop allele freq diff)"</span>)</div><div class='img'><img src='switchBOTgenotypes-2.png' alt='' width='700' height='432.632880098887' /></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Lyndal Henden.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>

